paddle/fluid/pybind/ir.cc

```diff
@@ -0,0 +1,112 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/pybind/ir.h"
+#include <string>
+#include <unordered_map>
+#include "paddle/fluid/framework/ir/graph.h"
+#include "paddle/fluid/framework/ir/node.h"
+#include "paddle/fluid/framework/op_desc.h"
+#include "paddle/fluid/framework/var_desc.h"
+#include "pybind11/stl.h"
+
+namespace py = pybind11;
+using paddle::framework::ir::Graph;
+using paddle::framework::ir::Node;
+using paddle::framework::OpDesc;
+using paddle::framework::ProgramDesc;
+using paddle::framework::VarDesc;
+using pybind11::return_value_policy;
+
+namespace paddle {
+namespace pybind {
+std::vector<Node *> GetNodesByType(const Node::Type &type, const Graph &graph);
+std::unordered_map<std::string, Node *> GetNodesMapByType(
+    const Node::Type &type, const Graph &graph);
+int Index(const Graph &graph, const Node &op);
+Node *Var(const Graph &graph, const std::string &name);
+Node *Var(const Graph &graph, const std::string &name);
+Node *CreateVar(Graph &graph, const std::string &name, const Node::Type &type,
+                const std::vector<int64_t> &shape);
+bool RemoveVar(Graph &graph, const std::string &name);
+bool RemoveOp(Graph &graph, int index);
+std::shared_ptr<Graph> Clone(Graph &graph);
+
+void BindGraph(py::module *m) {
+  py::class_<Graph, std::shared_ptr<Graph>>(*m, "Graph", "")
+      .def(py::init<const ProgramDesc &>())
+      .def("has", &Graph::Has)
+      .def("get_int", &Graph::Get<int>)
+      .def("get_float", &Graph::Get<float>)
+      .def("get_double", &Graph::Get<double>)
+      .def("get_string", &Graph::Get<std::string>)
+      .def("set", [](Graph &self, const std::string &attr_name,
+                     int attr) { return self.Set(attr_name, new int(attr)); })
+      .def("set",
+           [](Graph &self, const std::string &attr_name,
+              const std::string &attr) {
+             return self.Set(attr_name, new std::string(attr));
+           })
+      .def("set",
+           [](Graph &self, const std::string &attr_name, float attr) {
+             return self.Set(attr_name, new float(attr));
+           })
+      .def("set",
+           [](Graph &self, const std::string &attr_name, double attr) {
+             return self.Set(attr_name, new double(attr));
+           })
+      .def("erase", &Graph::Erase)
+      .def("nodes", &Graph::Nodes, return_value_policy::reference)
+      .def("create_var_node",
+           [](Graph &self, VarDesc &var_desc) -> std::shared_ptr<Node> {
+             return std::shared_ptr<Node>(self.CreateVarNode(&var_desc));
+           })
+      .def("create_op_node",
+           [](Graph &self, OpDesc &op_desc) -> std::shared_ptr<Node> {
+             return std::shared_ptr<Node>(self.CreateOpNode(&op_desc));
```

Comment:
Is shared_ptr required? Can return_value_policy::reference work?

Reply:
same above

Reply:
fixed.